### ISSUE TITLE
[Bug?] .env file not being read in worker mode

### ISSUE THREAD

**Issue by paullallier** [dunglas/frankenphp#308]
EDIT - It now looks like this is a .env file reading issue rather than a cache warmup one - I've fixed the warm-up problem but I'm still getting errors.

I'm currently moving my API-Platform dev / CI stack from Apache to FrankenPHP.  If I make a request to the API soon after the container starts, I typically get a response like this:
<img width="745" alt="Screenshot 2023-11-17 at 10 14 10" src="https://github.com/dunglas/frankenphp/assets/42591123/f2449a67-dabc-4b26-850a-45ed68aafeaa">

Most of the time, it eventually starts to work, so I assume it's Symfony cache warm up related?  (When I try run with a single worker, it never seems to warm up, but I suspect that's a config error by me.  I think I need single-worker mode for xdebug to be reliable?) In Apache, the early request just waits while the cache warms up and returns the response once it's finished (unless it times out).

I assume I'm seeing what you would expect?  So the questions are probably:

1. How can I warm up the cache correctly so that the container gets to a working state "instantly" on deployment.
2. How should I check that the warm up is complete?  It returns a 200 response even with that failure message.
3. And is there's a way to change things so that, at least in production, it can never leak that failure message even if the cache isn't warm?

This might be related to #220, but it feels a bit more Symfony and less Caddy deployment related than that.  It's also entirely possible that I'm simply using things wrong.  Once I understand what's going on I might be able to PR some docs to help others.

**Comment by paullallier** [comment]
Moving from #314, since that seems to have been a different issue.

It looks pretty consistently like it's not reading the .env file for me in worker model on my project.  Starting the container with `FRANKENPHP_CONFIG` undefined works.  But when I add `-e FRANKENPHP_CONFIG="worker ./public/index.php 1"` I get errors consistent with missing .env-defined variables.

Based on the 1.0.0-php8.3-bookworm image, using Symfony 7.0.1 and API Platform 3.2.7.  Running with:
```
ENV APP_RUNTIME=Runtime\\FrankenPhpSymfony\\Runtime
ENV CADDY_GENERAL_OPTIONS="auto_https off"
ENV SERVER_NAME="http://main-api"
ENV CADDY_GLOBAL_OPTIONS=debug
```

I have https turned off since I'm serving the site via an Nginx reverse proxy (in another container) which handles TLS and talked http to the FrankenPHP container.  But when I saw the issue I mentioned above, I was running it standalone with auto HTTPS on.

For legacy reasons, my app is being served from /var/www/main-api in the container rather than /app (I've changed the WORKDIR).  But if I copy the .env file to /app/ it still doesn't get read (checking that the worker isn't hard-coded to expect it in /app).   Defining them as "real" environment variables in the Dockerfile does work, so it looks like it's specifically something about reading .env files.

I'm not sure it's helpful, but I'm getting this in the FrankenPHP container logs:
```
{"level":"debug","ts":1702213534.1945422,"logger":"http.handlers.rewrite","msg":"rewrote request","request":{"remote_ip":"172.100.0.20","remote_port":"34208","client_ip":"172.100.0.20","proto":"HTTP/1.0","method":"GET","host":"main-api","uri":"/","headers":{"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7"],"X-Real-Ip":["172.100.0.1"],"X-Forwarded-Proto":["https"],"X-Nginx-Proxy":["true"],"Connection":["close"],"Sec-Ch-Ua-Mobile":["?0"],"Sec-Ch-Ua-Platform":["\"macOS\""],"Upgrade-Insecure-Requests":["1"],"Pragma":["no-cache"],"User-Agent":["Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"],"Sec-Fetch-Mode":["navigate"],"Cache-Control":["no-cache"],"Sec-Ch-Ua":["\"Not_A Brand\";v=\"8\", \"Chromium\";v=\"120\", \"Google Chrome\";v=\"120\""],"Sec-Fetch-Site":["none"],"Sec-Fetch-Dest":["document"],"Accept-Encoding":["gzip, deflate, br"],"Accept-Language":["en-GB,en-US;q=0.9,en;q=0.8"],"X-Forwarded-For":["172.100.0.1"],"X-Forwarded-Port":["443"],"Sec-Fetch-User":["?1"]}},"method":"GET","uri":"/index.php"}
{"level":"debug","ts":1702213534.1946356,"msg":"request handling started","worker":"/var/www/main-api/public/index.php","url":"/index.php"}
{"level":"debug","ts":1702213535.6216128,"msg":"request handling finished","url":"/index.php"}
{"level":"debug","ts":1702213535.6216726,"msg":"request handling finished","url":"/index.php"}
2023/12/10 13:05:35.621	ERROR	http.log.access.log0	handled request	{"request": {"remote_ip": "172.100.0.20", "remote_port": "34208", "client_ip": "172.100.0.20", "proto": "HTTP/1.0", "method": "GET", "host": "main-api", "uri": "/", "headers": {"Pragma": ["no-cache"], "User-Agent": ["Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"], "Sec-Fetch-Mode": ["navigate"], "Cache-Control": ["no-cache"], "Sec-Ch-Ua": ["\"Not_A Brand\";v=\"8\", \"Chromium\";v=\"120\", \"Google Chrome\";v=\"120\""], "Sec-Fetch-Site": ["none"], "Sec-Fetch-Dest": ["document"], "Accept-Encoding": ["gzip, deflate, br"], "Accept-Language": ["en-GB,en-US;q=0.9,en;q=0.8"], "X-Forwarded-For": ["172.100.0.1"], "X-Forwarded-Port": ["443"], "Sec-Fetch-User": ["?1"], "Accept": ["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7"], "X-Real-Ip": ["172.100.0.1"], "X-Forwarded-Proto": ["https"], "X-Nginx-Proxy": ["true"], "Connection": ["close"], "Sec-Ch-Ua-Mobile": ["?0"], "Sec-Ch-Ua-Platform": ["\"macOS\""], "Upgrade-Insecure-Requests": ["1"]}}, "bytes_read": 0, "user_id": "", "duration": 1.427297632, "size": 4902, "status": 500, "resp_headers": {"X-Debug-Token": ["55236e"], "Content-Type": ["application/problem+json; charset=utf-8"], "Vary": ["Accept"], "Content-Location": ["/errors/500"], "X-Debug-Token-Link": ["http://main-api/_profiler/55236e"], "Server": ["Caddy"], "X-Content-Type-Options": ["nosniff"], "Link": ["<http://www.w3.org/ns/hydra/error>; rel=\"http://www.w3.org/ns/json-ld#error\",<http://main-api/docs.jsonld>; rel=\"http://www.w3.org/ns/hydra/core#apiDocumentation\""], "X-Previous-Debug-Token": ["5d0937"], "Date": ["Sun, 10 Dec 2023 13:05:35 GMT"], "Pragma": ["no-cache"], "Expires": ["-1"], "X-Robots-Tag": ["noindex"], "X-Powered-By": ["PHP/8.3.0"], "X-Frame-Options": ["deny"], "Cache-Control": ["private, must-revalidate"]}}
{"level":"debug","ts":1702213535.8155124,"msg":"waiting for request","worker":"/var/www/main-api/public/index.php"}
```
(172.100.0.20 is the Nginx proxy, 172.100.0.1 is the IP address the containers see for the web browser on the host machine)

The browser response tells me that the worker is at least finding most of the files to serve:
<img width="583" alt="Screenshot 2023-12-10 at 13 15 59" src="https://github.com/dunglas/frankenphp/assets/42591123/c5fd4ad0-b5ae-4a1b-aa69-14d77867e93c">

If there isn't an obvious solution @dunglas - something I've mis-configured - I'll try to build a simple producer project.

EDIT: Oh and I do have `variables_order = EGPCS` in my phpinfo output

**Comment by paullallier** [comment]
I've been trying to dig into this some more and I'm still struggling.  I've tried running the demo project in as close to the same FrankenPHP config as my project, with all the same extensions, including trying (in both):

- serving HTTP and HTTPS;
- serving at localhost and another domain name;
- serving from the /app directory as the WORKDIR and from another directory;

The demo project consistently reads the .env file (I set the API title as an env variable to test) in worker mode.  And my project reads it in non-worker mode but never in worker mode.

I've run it with caddy debug turned on, but it's not telling me much.  Kevin, can you suggest anything else I can try to dig into the issue, please, since I've completely failed to make a simple reproducer?

Log output from my project (worker mode, http only, modified WORKDIR)
```2023-12-31 16:02:16 {"level":"warn","ts":1704038536.1842535,"logger":"http.auto_https","msg":"server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server","server_name":"srv0","http_port":80}
2023-12-31 16:02:16 {"level":"info","ts":1704038536.18503,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc000411a80"}
2023-12-31 16:02:16 {"level":"debug","ts":1704038536.184318,"logger":"http.auto_https","msg":"adjusted config","tls":{"automation":{"policies":[{}]}},"http":{"servers":{"srv0":{"listen":[":80"],"routes":[{"handle":[{"handler":"subroute","routes":[{"handle":[{"handler":"vars","root":"public/"},{"encodings":{"gzip":{},"zstd":{}},"handler":"encode","prefer":["zstd","gzip"]}]},{"handle":[{"handler":"static_response","headers":{"Location":["{http.request.orig_uri.path}/"]},"status_code":308}],"match":[{"file":{"try_files":["{http.request.uri.path}/index.php"]},"not":[{"path":["*/"]}]}]},{"handle":[{"handler":"rewrite","uri":"{http.matchers.file.relative}"}],"match":[{"file":{"split_path":[".php"],"try_files":["{http.request.uri.path}","{http.request.uri.path}/index.php","index.php"]}}]},{"handle":[{"handler":"php","split_path":[".php"]}],"match":[{"path":["*.php"]}]},{"handle":[{"handler":"file_server"}]}]}],"terminal":true}],"automatic_https":{"disable":true},"logs":{"logger_names":{"main-api":"log0"}}}}}}
2023-12-31 16:02:16 {"level":"debug","ts":1704038536.186272,"logger":"http","msg":"starting server loop","address":"[::]:80","tls":false,"http3":false}
2023-12-31 16:02:16 {"level":"info","ts":1704038536.1867058,"logger":"http.log","msg":"server running","name":"srv0","protocols":["h1","h2","h3"]}
2023-12-31 16:02:16 {"level":"debug","ts":1704038536.191244,"msg":"starting","worker":"/var/www/main-api/public/index.php"}
2023-12-31 16:02:16 {"level":"info","ts":1704038536.1937191,"logger":"tls","msg":"cleaning storage unit","storage":"FileStorage:/data/caddy"}
2023-12-31 16:02:16 {"level":"info","ts":1704038536.1943705,"logger":"tls","msg":"finished cleaning storage units"}
2023-12-31 16:02:16 {"level":"debug","ts":1704038536.36632,"msg":"waiting for request","worker":"/var/www/main-api/public/index.php"}
2023-12-31 16:02:16 {"level":"info","ts":1704038536.366437,"msg":"FrankenPHP started 🐘","php_version":"8.3.1"}
2023-12-31 16:02:16 {"level":"info","ts":1704038536.3673,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}
2023-12-31 16:02:16 {"level":"info","ts":1704038536.3673437,"msg":"serving initial configuration"}
2023-12-31 16:02:20 {"level":"debug","ts":1704038540.9129424,"logger":"http.handlers.rewrite","msg":"rewrote request","request":{"remote_ip":"192.168.65.1","remote_port":"39992","client_ip":"192.168.65.1","proto":"HTTP/1.1","method":"GET","host":"main-api","uri":"/","headers":{"User-Agent":["Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"],"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7"],"Accept-Encoding":["gzip, deflate"],"If-None-Match":["\"4fda40eadde691538290ce9597c93be5\""],"Connection":["keep-alive"],"Accept-Language":["en-GB,en;q=0.9"],"Cache-Control":["max-age=0"],"Upgrade-Insecure-Requests":["1"]}},"method":"GET","uri":"/index.php"}
2023-12-31 16:02:20 {"level":"debug","ts":1704038540.9132617,"msg":"request handling started","worker":"/var/www/main-api/public/index.php","url":"/index.php"}
2023-12-31 16:02:22 {"level":"debug","ts":1704038542.336652,"msg":"request handling finished","url":"/index.php"}
2023-12-31 16:02:22 {"level":"debug","ts":1704038542.3367305,"msg":"request handling finished","url":"/index.php"}
2023-12-31 16:02:22 2023/12/31 16:02:22.337     ERROR   http.log.access.log0    handled request {"request": {"remote_ip": "192.168.65.1", "remote_port": "39992", "client_ip": "192.168.65.1", "proto": "HTTP/1.1", "method": "GET", "host": "main-api", "uri": "/", "headers": {"User-Agent": ["Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"], "Accept": ["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7"], "Accept-Encoding": ["gzip, deflate"], "If-None-Match": ["\"4fda40eadde691538290ce9597c93be5\""], "Connection": ["keep-alive"], "Accept-Language": ["en-GB,en;q=0.9"], "Cache-Control": ["max-age=0"], "Upgrade-Insecure-Requests": ["1"]}}, "bytes_read": 0, "user_id": "", "duration": 1.424004133, "size": 4939, "status": 500, "resp_headers": {"X-Previous-Debug-Token": ["c3c1fd"], "Vary": ["Accept"], "X-Frame-Options": ["deny"], "Server": ["Caddy"], "X-Powered-By": ["PHP/8.3.1"], "X-Content-Type-Options": ["nosniff"], "Content-Location": ["/errors/500"], "Cache-Control": ["no-cache, private"], "Link": ["<http://www.w3.org/ns/hydra/error>; rel=\"http://www.w3.org/ns/json-ld#error\",<http://main-api/docs.jsonld>; rel=\"http://www.w3.org/ns/hydra/core#apiDocumentation\""], "X-Debug-Token": ["66579e"], "X-Robots-Tag": ["noindex"], "Content-Type": ["application/problem+json; charset=utf-8"], "Date": ["Sun, 31 Dec 2023 16:02:22 GMT"], "X-Debug-Token-Link": ["http://main-api/_profiler/66579e"]}}
2023-12-31 16:02:22 {"level":"debug","ts":1704038542.552087,"msg":"waiting for request","worker":"/var/www/main-api/public/index.php"}
```

**Comment by withinboredom** [comment]
This seems more like a symfony issue than a frankenphp issue (frankenphp doesn't read any `.env` files to my knowledge). `.env` files are an application level thing that needs support from the application. 

If you wanted to have Caddy/FrankenPHP load the env file and pass it on, I suspect [this might work](https://caddy.community/t/php-env-from-external-env-file/17657/2) since we do support the `env` directive (pretty sure anyway).

**Comment by paullallier** [comment]
Sorry Kévin - I just realised I would have pinged you when I copied my message from the other issue.  I was trying to avoid doing that on a weekend / holiday.

Thanks @withinboredom.  The fact that it works with Apache and with FrankenPHP in non-worker mode, but not in worker mode suggests to me that there's something FrankenPHP-specific going on.  But it's certainly possible it's a Symfony or (most likely) user configuration issue - especially since the demo project, which seems to work, is a Symfony/Api-Platform project just like mine.  I'll have a look at that --envfile option and see if it's at least a work around - thank you.
